### ISSUE TITLE
How to receive any "message" events?

### ISSUE THREAD

**Issue by DivineThreepwood** [haruntuncay/socket.io-client#1]
Hey haruntuncay,

I was already able to establish a connection to by socket.io cloud server and even the login was successful.
However,  I can not find any way to receive data messages. With the https://github.com/socketio/engine.io-client-java/blob/master/src/main/java/io/socket/engineio/client/Socket.java implementation, I was able to receive data messages by observing the` EVENT_MESSAGE = "message";` events but such event is not declared in your implementation. I already tried to observe these messages via the following code
<pre>
socket.on("message", objects -> {
                // handle request
                LOGGER.info("incoming message");
                handleRequest(objects[0], (Ack) objects[objects.length - 1]);
            });
</pre>
but the callback is never called even when the "message" packages are received by the client.
I couldn't find any related info in your documentation so could you offer any example code how to receive the default messages or do I get something totally wrong?

Cheers!
Marian

**Comment by haruntuncay** [comment]
Hi @DivineThreepwood , firstly, thank you for giving this implementation a shot :)., and sorry for the late response. This link https://github.com/socketio/engine.io-client-java/blob/master/src/main/java/io/socket/engineio/client/Socket.java  you've provided belongs to engine.io client, but the code sample you provided uses `ACK` objects, which are Socket.io features that aren't to be found in engine.io. Could you be able to share the way you create the  socket instance and how you send a message, please ? I will be more than happy to help.

**Comment by DivineThreepwood** [comment]
Sorry, the link was just wrong. The cloud server is implemented in JS and is using"socket.io": "2.2.0" and the old java client was using socket.io too:
<pre>
<dependency>
    <groupId>io.socket</groupId>
    <artifactId>socket.io-client</artifactId>
    <version>1.0.0</version>
</dependency>
</pre>

Actually, I am not maintaining the server code and have not much knowledge about JS but this must be the code section where the messages are send. https://github.com/openbase/bco.cloud/blob/cda5ae727ea934f95594a0794db73c7d5ed3fd55/socket_utils.js#L281
Here is the current client code were I created the socket:
<pre>
socket = IO.of(JPService.getProperty(JPCloudServerURI.class).getValue().toURL())
                        /* any options */
                        .header(ID_KEY, userId + "@" + bcoId)
                        .socket();
</pre>
old receiving code: https://github.com/openbase/bco.app/blob/8c243244b266553123bddade2d48b72b1d609a59/cloudconnector/src/main/java/org/openbase/bco/app/cloudconnector/SocketWrapper.java#L173 
new receiving code using your lib: https://github.com/openbase/bco.app/blob/2305015f1ec2dbdba98a4f55fd6694fc3494f8e5/cloudconnector/src/main/java/org/openbase/bco/app/cloudconnector/SocketWrapper.java#L164

**Comment by haruntuncay** [comment]
I am currently not having any problems receiving `"message"` events. Could you please try these 2 things:
1- Register for `Socket.ERROR` and `Socket.ABRUPT_CLOSE` events and see if either one is being called.
2- Could you be able to confirm that you are able to receive events other than `"message"` ?

Thank you in advance and sorry for the inconvenience.

**Comment by DivineThreepwood** [comment]
Hey,

> 1- Register for Socket.ERROR and Socket.ABRUPT_CLOSE events and see if either one is being called.
Both callbacks are not called when the "message" event is received.

> 2- Could you be able to confirm that you are able to receive events other than "message" ?
Currently, I have no access to the cloud service and therefore just "message" event are transmitted.
However, when adding a debug point in the Observer caller method, I can confirm that the data packet is received. The event is definitely passing the WebSocketTransport.java:100 line `emitEvent(Transport.PACKET, packet); `
<pre>
"Worker-Executor@4794" prio=5 tid=0x93 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at common.Observable.emitEvent(Observable.java:31)
	  at engineio_client.transports.WebSocketTransport.onPacket(WebSocketTransport.java:100)
	  at engineio_client.transports.WebSocketTransport.access$100(WebSocketTransport.java:13)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener.lambda$null$0(WebSocketTransport.java:122)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener$$Lambda$248.1190165464.call(Unknown Source:-1)
	  at engineio_client.parser.Parser.decodePacket(Parser.java:118)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener.lambda$onMessage$1(WebSocketTransport.java:122)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener$$Lambda$247.2039509025.run(Unknown Source:-1)
	  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	  at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:-1)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	  at java.lang.Thread.run(Thread.java:834)
</pre>
and is therefore correctly received on client site, right?

Many thanks for your help!

**Comment by DivineThreepwood** [comment]
The "message" event is then forwarded to the manager callback registered at  `Manager.java:109` -> `engine.on(EngineSocket.MESSAGE, this::onData);` and the onData method decodes the message string into a packet object and emit the packet `decoder.add((String) data, (packet) -> emitEvent(PACKET, packet));` which is why the message event is then never forwarded to the higher level client code, right?

**Comment by haruntuncay** [comment]
>  Manager.java:109 -> engine.on(EngineSocket.MESSAGE, this::onData); and the onData method decodes the message string into a packet object and emit the packet decoder.add((String) data, (packet) -> emitEvent(PACKET, packet));

This is correct but `emitEvent(PACKET, packet))` is actually has a callback registered by `Socket` like so: 
```
// socketio_client.Socket # line:109
callbackHandles.add(manager.on(Manager.PACKET, this::onPacket));
```

Which will then decode the packet into `eventName -> data` format and will emit whatever the `eventName` is. Which should call your callback. Could you please share with me how you got this output ?

> "Worker-Executor@4794" prio=5 tid=0x93 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at common.Observable.emitEvent(Observable.java:31)
	  at engineio_client.transports.WebSocketTransport.onPacket(WebSocketTransport.java:100)
	  at engineio_client.transports.WebSocketTransport.access$100(WebSocketTransport.java:13)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener.lambda$null$0(WebSocketTransport.java:122)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener$$Lambda$248.1190165464.call(Unknown Source:-1)
	  at engineio_client.parser.Parser.decodePacket(Parser.java:118)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener.lambda$onMessage$1(WebSocketTransport.java:122)
	  at engineio_client.transports.WebSocketTransport$WebSocketListener$$Lambda$247.2039509025.run(Unknown Source:-1)
	  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	  at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:-1)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	  at java.lang.Thread.run(Thread.java:834)

**Comment by DivineThreepwood** [comment]
Just by adding a debug breakpoint at `common.Observable.emitEvent(Observable.java:31)` and waiting until a message event is received.

**Comment by haruntuncay** [comment]
When you step into `common.Observable.emitEvent(Observable.java:31)`, doesn't it call any callbacks ?
Because that's exactly when the `Socket.onPacket` should be called which will call your callback in return.

**Comment by DivineThreepwood** [comment]
okay we are getting closer to the issue. EngineSocket.java:283 `emitEvent(MESSAGE, packet.getData());` is called but not my callback.

**Comment by haruntuncay** [comment]
Sorry that you're having this struggule with this lib, I have to add logging immediately in 2 or 3 days so if this kind of error arises ever again, we would just have to look at the logs instead of dealing with it manually. Could you be able to share the contents of `packet.getData`  ?

**Comment by DivineThreepwood** [comment]
No worries :) I am just happy to switch to a maintaint socket.io implementation :)
<pre>
packet = {222["message","{\"inputs\":[{\"context\":{\"locale_country\":\"DE\",\"locale_language\":\"de\"},\"intent\":\"action.devices.EXECUTE\",\"payload\":{\"commands\":[{\"devices\":[{\"id\":\"53b59c91-dd89-4a24-95ae-0ba841634039\"},{\"id\":\"a27ba29b-6949-417a-8090-13354173357b\"},{\"id\":\"9deaa7ec-5608-4f56-ab16-4f732c41429c\"},{\"id\":\"cf2724dd-9657-4a81-b4e6-6038d4bafbab\"},{\"id\":\"b968f74e-0c3d-46af-8b26-b99810a7c6e1\"},{\"id\":\"93f2f1d8-d003-4f83-a397-5e5747894124\"},{\"id\":\"a0f2c9d8-41a6-45c6-9609-5686b6733d4e\"},{\"id\":\"d54b7e30-5ab0-4bfa-992d-1fc154a2b886\"},{\"id\":\"cdb80a3d-e2bf-47c8-a7a8-9bb2c07ea5b6\"},{\"id\":\"67e7ca76-93db-46e2-b129-64998265ef8f\"},{\"id\":\"34ec31ee-dc02-4377-88d7-2eb705ed041f\"},{\"id\":\"de88f2f8-84f2-4b7c-9f41-aa8eee6aa805\"},{\"id\":\"fb771e44-5ba3-4ba7-95b9-63a6ee0cf6f6\"},{\"id\":\"5b0062a8-b76b-474a-bd6b-3ba67676161d\"},{\"id\":\"f5a8f27a-590c-4074-8f01-bbd34634db95\"},{\"id\":\"d3f1c006-131a-431f-b2d7-d68bdd056548\"},{\"id\":\"99aec777-6fa4-403b-ae01-ce78754296a3\"},{\"id\":\"0bfcd9ee-3e5b-415a-b7fd-79bb3ff45bb3\"},{\"id\":\"41b6d59d-e47a-4868-b770-b52654e9b69d\"},{\"id\":\"f01734d6-08b8-4d30-9581-d6ad5cbfe839\"},{\"id\":\"2095fe78-3e25-4510-a053-3b353ea9599b\"},{\"id\":\"1999d5f9-9a68-48c9-919d-80bb2038e4b6\"},{\"id\":\"9b32481b-6a0a-4e21-9306-7d6956b4012f\"},{\"id\":\"b178d3ac-b1cf-4f9a-8ba3-13796039721c\"},{\"id\":\"96497d9f-87fa-462b-b599-0b49be855983\"},{\"id\":\"82dc47b8-e590-4b4c-9019-745f99800429\"},{\"id\":\"852ced01-db8f-4ad8-8489-3d0c3158afb0\"},{\"id\":\"f65262e0-f7b2-4f71-a540-db1179fdb90e\"},{\"id\":\"0e11326e-f2fb-4b61-80d3-51f36b69ea22\"},{\"id\":\"24b6a4f7-dadc-4d44-a8c8-55d8e7ac7e94\"},{\"id\":\"36fd7e2c-6e82-4d79-9f49-ab78795a9829\"},{\"id\":\"d5c209f2-322f-45e9-87c9-97381c6d8188\"},{\"id\":\"af092f4b-a8f0-45cb-b205-06b0f0af93fb\"},{\"id\":\"ecd6342d-78e5-4c00-a5ca-ca6364066efe\"},{\"id\":\"6440edf5-0af7-4799-8b92-0471ca736e63\"}],\"execution\":[{\"command\":\"action.devices.commands.OnOff\",\"params\":{\"on\":true}}]}]}}],\"requestId\":\"15116617931836343442\"}"]}
</pre>

**Comment by haruntuncay** [comment]
Thank you for the help and sorry that you've to deal with a cumbersome issue. Could you please put a break point at `socketio_client.Socket#onPacket (line:196)` and see whether it calls the `handleEventPacket`, or not ?
Thanks in advance.

**Comment by DivineThreepwood** [comment]
Sure, I just placed the break point at `socketio_client.Socket:196` but the breakpoint is never passed when the "message" event is received.

**Comment by haruntuncay** [comment]
Hello Marian,
Sorry for the late response. I've added simple logging to socket.io 1.0, so if you remove it from your maven repo cache and install again, it should be able to use logging. But your problem most likely caused due to a connection error. Could you register listeners for `Socket.ERROR`, `Socket.ABRUPT_CLOSE`, `Socket.CLOSE` and `Socket.DISCONNECT` and see which one is getting called ?
Thanks in advance.

**Comment by DivineThreepwood** [comment]
Hey Harun, I just locally build your master so you don't need to patch any releases.
I already registered all kind of error callbacks but non of them are called: https://github.com/openbase/bco.app/blob/feature-switch-socketio-impl/cloudconnector/src/main/java/org/openbase/bco/app/cloudconnector/SocketWrapper.java#L203
However, after you added the logging the following error is printed:
<pre>
02:10:38.858 [OkHttp https://bco-cloud.herokuapp.com/...] ERROR e.t.PollingTransport Response error during poll request. Response: Response{protocol=http/1.1, code=400, message=Bad Request, url=https://bco-cloud.herokuapp.com/socket.io/?EIO=3&sid=WOLi_mXmTSc0FQb9AAAC&transport=polling}
</pre>

After adding some more logging I can confirm that the "message" event is processed by Manager.java:158 (onData(Object... args)) but is never passing the Socket.java:195 (onPacket(...)) method because the decodeStringData() of ParserImpl.java throws the following NumberFormatException:
<pre>
java.lang.NumberFormatException: For input string: "4["message","{\"inputs\":[{\"context\":{\"locale_country\":\"DE\",\"locale_language\":\"de\"},\"intent\":\"action.devices.EXECUTE\",\"payload\":{\"commands\":[{\"devices\":[{\"id\":\"a687b6d9"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at socketio_client.parser.ParserImpl$DecoderImpl.decodeStringData(ParserImpl.java:193)
	at socketio_client.parser.ParserImpl$DecoderImpl.add(ParserImpl.java:125)
	at socketio_client.Manager.lambda$onData$10(Manager.java:165)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

</pre>

**Comment by DivineThreepwood** [comment]
Is there something wrong with the message that I received from my cloud service or is this just a bug in  in your decoding phase? At saturday I have a coding session with my cloud developer and I would like to address this issue in case its caused by our cloud service. So I would really appreciate to get a little hint from your side :)

**Comment by haruntuncay** [comment]
@DivineThreepwood, really sorry for the late response, and thank you for the logs and the issue. Turns out it was a bug, as you also stated. I have submitted a commit fixing it, and have updated the 1.0 repository. If you again clear the cached version from your local maven repo and install again, it should be fixed.

**Comment by haruntuncay** [comment]
@DivineThreepwood, it seems like there is an issue with maven central, if the updated versin of 1.0 doesn't work for you, please let me known so that I can try to release 1.1 instead.

**Comment by DivineThreepwood** [comment]
Could you explain what you mean with "updated the 1.0 repository". As long as I know, your are not able to replace an already already uploaded artifact at central maven which is anyway not a good idea at all. Why not just releasing a patched version 1.0.1? There is not yet any update of your lib listed since the 8 of feb. https://repo1.maven.org/maven2/com/github/haruntuncay/socket.io-client/

**Comment by DivineThreepwood** [comment]
Hey Harun,
I locally build your master branch and I can confirm that the issue is fixed by commit: 2e86052
socket.id client is now working like expected :) So after you released a 1.0.1 patch I will start testing it in my nightly releases.
Thanks a lot for your support and your spend effort to implement this library.
